test/form-builder/TextField.test.tsx

```diff
@@ -1,17 +1,83 @@
+/* eslint-disable @typescript-eslint/ban-ts-comment */
+/* eslint-disable @typescript-eslint/no-empty-function */
+
 import React from 'react';
 import TextField from '../../src/form-builder/TextField';
+import { waitFor } from '@testing-library/react';
+
+import { buildRenderForm, changeValue } from '../utils';
 
-import { buildRenderForm } from '../utils';
+const renderForm = buildRenderForm();
 
-const renderForm = buildRenderForm({ name: 'asdf' });
+const getInput = (container: HTMLElement) => {
+  const input = container.querySelector('va-text-input') as HTMLInputElement;
+  if (!input) throw new Error('No va-text-input found');
+  return input;
+};
 
 describe('Form Builder - TextField', () => {
   test('renders', () => {
-    const { getByLabelText } = renderForm(
+    const { container } = renderForm(
+      <TextField name="thing" label="The Thing" />
+    );
+    const input = getInput(container);
+    expect(input.getAttribute('label')).toEqual('The Thing');
+    expect(input.getAttribute('name')).toEqual('thing');
+  });
+
+  test('renders initial value', () => {
+    const rf = buildRenderForm({ thing: 'asdf' });
+    const { container } = rf(<TextField name="thing" label="The Thing" />);
+    const input = getInput(container);
+    expect(input.getAttribute('value')).toEqual('asdf');
+  });
+
+  test('renders the default "required" validation error message', async () => {
+    const { container, getFormProps } = renderForm(
+      <TextField name="thing" label="The Thing" required />
+    );
+    const input = getInput(container);
+    await waitFor(() => {
+      getFormProps().setFieldTouched('thing');
+    });
+    expect(input?.getAttribute('error')).toEqual('Please provide a response');
```

Comment:
I was wondering why we have the optional chaining here, but then I realized it's because `getInput` might throw an error instead of return, and so Typescript would complain about it otherwise. I don't love that in this case because for this test we _know_ that there will be an input despite what the operator suggests. However, a `null` value still wouldn't equal the error message so at least we won't get false positives... :shrug: 

Would we be better off using the non-null assertion operator (`!`) here? I can see that error message having some value for debugging and writing initial tests, but beyond that its value goes down. Would a log message be better than throwing an error?

Reply:
Right, we won't run into false positives. :man_shrugging: 

As for the non-null assertion, that might work fine, but what would we gain from it?

> Would a log message be better than throwing an error?

I'm not sure I understand this. What do you mean?

Reply:
`getInput` has:

```js
if (!input) throw new Error('No va-text-input found');
```

Would it be better to have something like `console.error()` instead?

Reply:
Regarding what we gain by the non-null assertion operator, I would argue clarity. When I looked at this test it was confusing, but using the non-null combined with a reader who knows what it means, I think this would be easier to understand.

Reply:
> Would it be better to have something like `console.error()` instead?

No. If we wrote to `stderr`, it wouldn't affect the test itself, so the test would fail with a `cannot get property getAttribute of undefined` error, and we'd have to scroll up in the test output to find the `console.error()` message. By throwing an error, the test runner (jest) knows to display that error message when the test fails.

> Regarding what we gain by the non-null assertion operator, I would argue clarity. When I looked at this test it was confusing, but using the non-null combined with a reader who knows what it means, I think this would be easier to understand.

That's fair. I actually just added a return type to `getInput` and that seemed to eliminate the need for both optional chaining and the non-null assertion. :man_shrugging: I think because previously, the implied type was `HTMLInputElement | undefined` (if it threw the error).